zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZNodeGroupAclProperties.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.apache.zookeeper.server.auth.znode.groupacl;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Configured properties for ZNode Group ACL feature
+ */
+public class ZNodeGroupAclProperties {
+  private static ZNodeGroupAclProperties instance = null;
+
+  private ZNodeGroupAclProperties() {
+  }
+
+  public static ZNodeGroupAclProperties getInstance() {
+    if (instance == null) {
+      synchronized (ZNodeGroupAclProperties.class) {
+        if (instance == null) {
+          instance = new ZNodeGroupAclProperties();
+        }
+      }
+    }
+    return instance;
+  }
+
+  /**
+   * Property key values (JVM configs) for ZNode group ACL features
+   */
+  public static final String ZNODE_GROUP_ACL_CONFIG_PREFIX = "zookeeper.ssl.znodeGroupAcl.";
+  // Enables/disables whether znodes created by auth'ed clients
+  // should have ACL fields populated with the client Id given by the authentication provider.
+  // Has the same effect as the ZK client using ZooDefs.Ids.CREATOR_ALL_ACL.
+  public static final String SET_X509_CLIENT_ID_AS_ACL =
+      ZNODE_GROUP_ACL_CONFIG_PREFIX + "setX509ClientIdAsAcl";
+  // A list of domain names that will have super user privilege, separated by ","
+  private static final String SUPER_USER_DOMAIN_NAME =
+      ZNODE_GROUP_ACL_CONFIG_PREFIX + "superUserDomainName";
+  // A list of znode path prefixes, separated by ","
+  // Znode whose path starts with the defined path prefix would have open read access
+  // Meaning the znode will have (world:anyone, r) ACL
+  private static final String OPEN_READ_ACCESS_PATH_PREFIX =
+      ZNODE_GROUP_ACL_CONFIG_PREFIX + "openReadAccessPathPrefix";
+  private Set<String> openReadAccessPathPrefixes;
+  private Set<String> superUserDomainNames;
```

Comment:
`private volatile`

Reply:
The sets are not going to be changed after initialization, because the properties will be only loaded once. Also the initialization is already synchronized. So I think it's not necessary to use `volatile` here.

Reply:
Regarding `volatile` - using volatile can prevent creating multiple objects which is unlikely, but I have a concern on the overhead of `volatile` on the performance of read the object. Discussed offline with @narendly, decided not to use `volatile`.